### ISSUE TITLE
How to do the initial data transfer of the basis data (first data base before users add more)

### ISSUE THREAD

**Issue by jac2130** [CollectiWise/owlab#8]
Here I will make some suggestions and statements about what definitely should be the case for the first issue in integrating the two major parts of CollectiWise which each other. @romanradin  There are two different elements to integration between the ConceptNet part of CollectiWise and the Opencog part of CollectiWise.   The first one is the initial data _before there are any users_ and then there is the part where users enter new data, which is really different!  

For the **first part** there isn't any need for Implications; only Concepts, Relations, Relationships, Predicates and Attributes and it should be in this order.  The second part when users are adding data has to be in real time because users must get immediate feedback to what they are doing. One of the most relevant functions is `lookup_cases` but also the fact that when they make a new statement the points for making that statement must be immediately deducted and **points are handled withing the Opencog part of the system**.    

For loading in the initial data, computer scientists call this **the Base [Ontology](https://artint.info/html/ArtInt_316.html)** so I will refer to the initial data as the base ontology and to what the users are putting in as the **user defined statements**.  I hope that this won't cause confusion!  

It is also very important that the Base Ontology can be a different data set, such as [this one](http://aims.fao.org/activity/blog/sustainable-development-goals-interface-ontology-sdgio-support-united-nations) which interests the UN or another completely different one, for example [this one](http://current.geneontology.org/ontology/go-basic.json) ; the tables are going to be the same though.  

So for sure, everything should be done on the Google Cloud and we should use whatever Google Cloud services that are best suited for the given task.  The first task is to run the scripts on some Google service that is built in particular for the purpose to transfer massive data (note that the base ontology we are using is small compared to what it potentially could be!) from one relational data base to the AtomSpace via our API.  This should definitely be done using the most cutting edge tools that Google provides and not using something that works but takes forever to do; that is precisely the point of using Google services and we must make use of their best stuff!  To begin with, both APIs should be on Docker containers and they must definitely be on a Kubernetes orchestration system; that was part of Scope one of this project and this will most definitely be necessary for it to be a scalable system and that it must be; no serious customer like the UN, which I'm trying to get will forgive it not being scalable or slow in any way or failing because it isn't fault line tolerant; this is an absolute must of this particular system.  Then, we need to use some method that Google has built to do the first integration task which is the base line ontology transfer from the ConceptNet database, which should be in [Cloud Spanner ](https://cloud.google.com/spanner/?utm_source=google&utm_medium=cpc&utm_campaign=emea-gb-all-en-dr-skws-all-all-trial-e-gcp-1008073&utm_content=text-ad-none-any-DEV_c-CRE_253486129176-ADGP_Hybrid+%7C+AW+SEM+%7C+SKWS+~+EXA_M:1_GB_EN_Spanner_Home_cloud+spanner-KWID_43700053283788720-kwd-283006412836-userloc_9045999&utm_term=KW_cloud%20spanner-NET_g-PLAC_&ds_rl=1242850&ds_rl=1245734&gclid=Cj0KCQjws_r0BRCwARIsAMxfDRid9CA7Eb7gYeVfnDRMf9VxxXA1z4Kx0eszYogkAOguiaHztrjXxskaAm7zEALw_wcB) to the AtomSpace, which should also be persistent on Cloud Spanner (Cloud Spanner is definitely the right data base service as I've studied this issue for some time now and it is optimized for the use cases that we need).  We need definitely to run these scripts in parallel and I think but I'm not sure that [PySpark and DataProc](https://linuxacademy.com/hands-on-lab/a2e11476-3b13-4164-a6da-288935c1e790/) are the right solution.  Another possible solution involves [DataFlow](https://cloud.google.com/dataflow/?utm_source=google&utm_medium=cpc&utm_campaign=emea-gb-all-en-dr-skws-all-all-trial-e-gcp-1008073&utm_content=text-ad-none-any-DEV_c-CRE_253486130814-ADGP_Hybrid+%7C+AW+SEM+%7C+SKWS+~+EXA_1:1_GB_EN_Cloud_Data_Dataflow_Gap+Analysis+01.11.2017-KWID_43700053279204007-kwd-301040544953-userloc_9045999&utm_term=KW_data%20flow-NET_g-PLAC_&ds_rl=1242850&ds_rl=1245734&gclid=Cj0KCQjws_r0BRCwARIsAMxfDRjx4vjBv5qLGSK2eAV9w75bU7_WZiYzHNcikYSm5dIkN8tuR5O4l6gaAoDwEALw_wcB) ...there might be some other Google Cloud services that are better suited for this task but it must be built for massive data transfers using API calls on parallel using the API on Kubernetes.

**Comment by romanradin** [comment]
will discuss this with the dev a little later

**Comment by jac2130** [comment]
[This](https://opensource.googleblog.com/2020/02/harbourbridge-from-postgresql-to-cloud.html) might be relevant!

**Comment by jac2130** [comment]
Also this: 
https://github.com/gregsramblings/google-cloud-4-words/blob/master/DarkBrochure.pdf

**Comment by romanradin** [comment]
@jac2130 
our initial plan was to use the native django tools as an interlayer between SQL DB & the mechanism responsible for queries sending. these tools are very cool & powerful. they are optimized to work with big databases. if we use django tools all we will need is a usual virtual machine with a powerful processor.  
now we're told to use clusters. google clusters use the Hadoop system that doesn't interact with the operating system. they interact with Haddop only using such tools as Apache spark. Apache spark is a big framework to work with big data. Haddop & Apache spark are not the tools that our developer uses every day. that's why he needs the time to discover them. 
so, the situation is: the functions that we created before won't work with clusters. so, they need to be created almost from scratch. 

question1: is it really worthy to use clusters for a one time job? 

to use the virtual machine variant we just need 3 things:
- set up Atomspace  
- set up a virtual machine in google cloud environment 
- run it & wait untill its job is done.

question2: why not to manage this Atomspace problem that we're having for 2-3 weeks?

**Comment by romanradin** [comment]
>For the first part there isn't any need for Implications; only Concepts, Relations, Relationships, Predicates and Attributes and it should be in this order.

we have the functions to move this data already. but if you insist on using clusters then we have to create them almost from scratch

**Comment by romanradin** [comment]
>The second part when users are adding data has to be in real time because users must get immediate feedback to what they are doing. One of the most relevant functions is lookup_cases but also the fact that when they make a new statement the points for making that statement must be immediately deducted and points are handled withing the Opencog part of the system.

this can't be done still as we have the problems with opencog

**Comment by jac2130** [comment]
> question1: is it really worthy to use clusters for a one time job?

Well, so one of the features that are important is merging of **Ontologies** as in [this tool, named pronto](https://github.com/jac2130/pronto) but also with [JSON-LD](https://json-ld.org/) not just with [owl](https://www.w3.org/OWL/), [obo](http://www.obofoundry.org/) and [xml](https://en.wikipedia.org/wiki/XML) with bigger data and including Opencog.  

This means that it really isn't a one time thing: it must be possible to merge ontologies (meaning that it must be possible to feed in more big data later on) and that this ought to go pretty fast.  If you play around a bit with pronto you see what I mean. 

Remember that the problem is that **it will take 25000 hours** (maybe a bit less with Google Compute Engine but not much less, in the way that you've built it?  That would be a serious problem.  

The other thing that I can do is that you make it work now with Django somehow and Google Compute Engine and that I later hire someone to rewrite everything in PySpark. Or I do it myself with Scala Spark **which is actually better** (because Spark is native to Scala and Scala is fully functional) but no one knows Scala; in that case I'll just follow your scrips and copy them into Scala Spark and that is what I did at eBay anyway.  So I could do that.  But then your solution must work faster than 25000 hours somehow?   

> to use the virtual machine variant we just need 3 things: 
>     * set up Atomspace
> 
>     * set up a virtual machine in google cloud environment
> 
>     * run it & wait untill its job is done.
> 
> 
> question2: why not to manage this Atomspace problem that we're having for 2-3 weeks?


I haven't had a chance yet because I'm trying to get money as much as possible and as fast as possible so I have applied for a loan where I need to do a lot of work with budget forecasting etc and that is still not done.  I have already applied for the grant but I won't hear anything back until sometime in May about that and I'm trying to get money faster than that; like withing two or three weeks time ...so I still have to do work there. Unfortunately I don't have anyone doing any of that for me yet.  There are also some contracts that I'd like to get (one with the UN and some other ones with the US government) but for that I need time and also a demonstrable system ...but I first need time and call them a bunch of times etc ...so in short, I haven't had the time to solve that AtomSpace problem and I don't yet have the cash to have someone else solve it. For some reason I haven't heard anything back from Nil and the opencog people in a while either.

**Comment by romanradin** [comment]
> Remember that the problem is that it will take 25000 hours in the way that you've built it?

seems like you didn't understand us. 2500 hour was the approx ETA for the case in which we use local laptop instead of cloud services. our current idea is to use a virtual machine in google cloud environment

**Comment by jac2130** [comment]
> > Remember that the problem is that it will take 25000 hours in the way that you've built it?
> 
> seems like you didn't understand us. 2500 hour was the approx ETA for the case in which we use local laptop instead of cloud services. our current idea is to use a virtual machine in google cloud environment

OK can you tell me how long that would take, approximately?

**Comment by romanradin** [comment]
>OK can you tell me how long that would take, approximately?

the scheme that includes Google compute engine +  our Python scripts will make this data transfer job in not more than a week. this is a very approximate ETA

**Comment by romanradin** [comment]
> I haven't had the time to solve that AtomSpace problem and I don't yet have the cash to have someone else solve it.

anyway you have to remember that your system won't be fully working system without a fully working Atomspace

**Comment by jac2130** [comment]
A week is pretty long still but sure ...lets try this!  In the meantime, could you send me those scripts and I'll play around with them whenever I have time and see if it would take me very long to rewrite them in Spark?

**Comment by romanradin** [comment]
ok. will do

**Comment by romanradin** [comment]
so, this becomes a priority 
![image](https://user-images.githubusercontent.com/13418239/79985842-db364200-84b3-11ea-9242-fcee07e2b5d6.png)

since you decided that we can work according to our scheme

**Comment by jac2130** [comment]
yes indeed ...if it won't take me too long to rewrite your script it might make sense for me to do those two things in one go though

**Comment by jac2130** [comment]
it'd be better to run this all in DataProc and see if we can do this quickly (run the whole thing in under one hour)

**Comment by romanradin** [comment]
>it'd be better to run this all in DataProc and see if we can do this quickly (run the whole thing in under one hour)

have you seen this? 
![image](https://user-images.githubusercontent.com/13418239/79986393-a70f5100-84b4-11ea-8f01-363def840696.png)

so, you told us that we won't use clusters for this job

**Comment by jac2130** [comment]
yes but if I could do it quickly maybe at least I should try real quick ...if it looks prohibitively complex then I agree that we should do it this way first

**Comment by jac2130** [comment]
if it takes one week to run then maybe working on it for 5 hours to run it under one hour is worth it?

**Comment by jac2130** [comment]
...just a thought

**Comment by romanradin** [comment]
let me remind you what's the current Atomspace problem:
- Atomspace keeps the data in RAM. so, if the server is reboted, for example, the data will be lost. so SQL needs to be connected to Atomspace so that the data would be transferred back & forth in real time & the data wouldn't be lost 
- all of our attempts to set up the interaction between Atomspace & SQL failed (according to the manual provided by Atomspace guys)
- so, currently Atomspace works well only till the moment the server goes done. this will be the moment the whole data will be lost. 
- summary: whatever we do, the final system will loose all the data some day. so, managing the problem of (Atomspace + persistent SQL DB) interaction has to be the highest priority , we think

**Comment by jac2130** [comment]
One week means that I can't demo it that quickly ...I'd like to show the working system to some people

**Comment by jac2130** [comment]
> One week means that I can't demo it that quickly ...I'd like to show the working system to some people

Definitely no matter how we run this thing, we must first solve the AtomSpace issue!  I agree with you on that! 100%

**Comment by romanradin** [comment]
remaking our functions to work with clusters may take more than one week

**Comment by jac2130** [comment]
All I'm saying is that if it is super easy for me to rewrite the code in Spark I should do both, fix the AtomSpace issue AND rewrite the thing

**Comment by jac2130** [comment]
and then run the thing

**Comment by jac2130** [comment]
> remaking our functions to work with clusters may take more than one week

Maybe and if it looks that way I will not do it now; but I'm pretty fast in Scala coding

**Comment by romanradin** [comment]
https://www.dropbox.com/s/li4gxihdh8lp6bs/commands.rar?dl=0 here are our functions

**Comment by jac2130** [comment]
Thank you!  Ok so today I have a call at 3PM about that loan and depending on how much work I have to do on that loan, I'll start working on the AtomSpace issue right away ...then I'll check quickly if it looks like a lot or a little work to rewrite this code in Scala and depending on that I'll let you know.  In the meantime, it would help to work on the real time user interactions with the Opencog system and for that, regular python (Django) code will definitely work; that way we can work in parallel?

**Comment by jac2130** [comment]
Maybe it would help to discuss the next step which includes users to call the API from the front-end?

**Comment by romanradin** [comment]
>Maybe it would help to discuss the next step which includes users to call the API from the front-end?

it's possible that smth may break on our side if we try to do the next steps before the interaction between the scripts & Atomspace works well

**Comment by jac2130** [comment]
From our Slack discussion; I don't understand this:
> 4. Run such command using the previously copied path to the container https://prnt.sc/rva6ag

Does it have to do with attaching volumes?  How do you connect the container to a database and what database?  I know that when you set up things you have to reach out of the container to a persistent database; where are you doing it and how?  Note: I'm not using this container (https://prnt.sc/rva6ag);  I'm using the one I made and that I know worked and then I'm cloning the `owlwise` git repo. I'm following the last path of the options you told me about; using the image that I've built myself so that I can go back and rebuild and I know it works (gcr.io/collectiwise/flask-oc@sha256:24be2ed9ad1bb11e846894c3a1c7bcd51e1cb317313932c13bf53b8d632f603c):

> Clone the repo to the environment with a working OpenCog system and run these commands:
> `cd owlwise`
> `pip install -r requirements.txt`
> `python manage.py runserver` 

So, I've tried getting the database Scheme error and I didn't get that error at all.  But I don't know how to probe this thing at all ...I can't tell whether it is attached to a persistent database or not.  

![DatabaseAttachingNoErrors](https://user-images.githubusercontent.com/1983503/80513830-8a24c300-8977-11ea-8b08-e2a846c98d0a.png)

**Comment by jac2130** [comment]
Do I need to set up a volume ...how do I connect this sort of database to the docker container? 

![CloudSpanner](https://user-images.githubusercontent.com/1983503/80514168-0fa87300-8978-11ea-92e5-16892db38c8d.png)

**Comment by jac2130** [comment]
I ndon't really know how this Django server works or how to probe it to see if I can get the same error you were getting: 
![djangoManage](https://user-images.githubusercontent.com/1983503/80517112-7465cc80-897c-11ea-88ee-b8150f40edbb.png)

**Comment by jac2130** [comment]
So now I'm getting this error which I shall be eliminating since it happens in my code: 
![alphabet_cycle_error](https://user-images.githubusercontent.com/1983503/80520842-eee51b00-8981-11ea-82ba-4abd4fa046ec.png)

**Comment by jac2130** [comment]
So now it's been on the following screen for some time! 
![waitingToConnect](https://user-images.githubusercontent.com/1983503/80521809-8bf48380-8983-11ea-92db-7379c5c3df39.png)

**Comment by jac2130** [comment]
Okay so now I have un-commented the problem code and I still didn't get the error; I got some other strange behavior: 
![RelationExistsSkipping](https://user-images.githubusercontent.com/1983503/80524584-c102d500-8987-11ea-8b25-7c4cb3782128.png)

**Comment by jac2130** [comment]
@romanradin do you see the last message above?  What am I not doing?  I'm not getting the error that you got!

**Comment by romanradin** [comment]
>Okay so now I have un-commented the problem code and I still didn't get the error; I got some other strange behavior:

the key error here is that opencog functions don't let the django thread start

**Comment by romanradin** [comment]
>Okay so now I have un-commented the problem code and I still didn't get the error; I got some other strange behavior:

@jac2130 the key error here is that opencog functions don't let the django thread start. because after the things we see at your last screenshot we had to see some django logs, but we don't see them

**Comment by jac2130** [comment]
So how do I see the error message for that @romanradin ?  I need to see where and how it fails so that I can go to the offending code and fix the problem!  This should not fail quietly ...there must be an informative error message?

**Comment by jac2130** [comment]
OK so there is interference between Django and the SQL done in Scheme.  What database are we trying to connect to?  And are we trying to connect to the database using Django?!  That would make things certainly slow! @romanradin

**Comment by jac2130** [comment]
This is the picture we _really_ need:
Basically we can follow the example of the script that Vitaliy created in Scope 1 and mount a volume for OpenCog the same way that it is done for ConceptNet in the Kubernetes code.  Then we have a Kubernetes Cluster with _K_ worker nodes for the OpenCog API which then are all mounted to the same volume database, which should simply be a Google Cloud Spanner instance.  
![CNtoOP](https://user-images.githubusercontent.com/1983503/80712451-ec9dd080-8ae9-11ea-9651-e12131a8f004.JPG)

**Comment by jac2130** [comment]
The most important part @romanradin is that a container should talk with the frontend and with the source data via a Django API, that is perfectly fine, but it should **not ever** talk to its sink database using Django!  It should instead use volume mounting.  https://docs.docker.com/storage/volumes/  ...this is very important and might be the reason the code is breaking currently ...there should be no problem with using SQL otherwise.

**Comment by romanradin** [comment]
@jac2130 I need to discuss all your comments from this card with Vitaliy & our backend developer

**Comment by jac2130** [comment]
OK sounds great; lets get the ball rolling!

**Comment by jac2130** [comment]
Here is an old system diagram as to how things should work: 

![SystemDiagram](https://user-images.githubusercontent.com/1983503/80941597-c1adc800-8dda-11ea-8620-f19d2e1a47a3.jpg)
